src/generic/AbsSeries.jl

```diff
@@ -134,6 +134,10 @@ function Base.hash(a::AbstractAlgebra.AbsSeriesElem, h::UInt)
    return b
 end
 
+function characteristic(a::AbsSeriesRing{T}) where T <: RingElement
```

Comment:
The `where T <: RingElement` can be removed, it just adds noise. The `where T` construct should in general be used only for dispatch purposes, to restrict the accepted types, which is not the case here (and then, the convention is to write e.g.  `characteristic(a::AbsSeriesRing{<:SomeRings})` if `T` is not used within the function body.)

Reply:
It also serves as implicit documentation though.

Reply:
The type shouldn't be Any here. 

Are you absolutely sure the variant you give is equivalent? I thought that was now used for specifying covariance, whereas "where" is for a unionall or some such thing.

I've been using "where" as it seems to be correct for the cases we use it.

I'm not sure if the two different notations would make any real difference here. But I think that conceptually they are different.

Reply:
You can see it like that, but `where` was never meant for documentation. In this function, I don't need to receive the information that `AbsSeriesRing` is defined only for `RingElement`. A `where` clause in a function catches the attention of the reader, it signifies: I will only work for these specific `T`s, it's a filter for dispatch. To add a filter which filters nothing out is a useless load on the reader. In addition, it requires more maintenance, as now it must be updated if `AbsSeries` is modified to accept more types.

Reply:
> Are you absolutely sure the variant you give is equivalent?

In this case yes, as `AbsSeriesRing{T}` is defined only for `T <: RingElement`. I don't really understand your comment about covariance.

> But I think that conceptually they are different.

They are somewhat conceptually different, as I said in my previous comment: a `where` in a function is for dispatch only, so by putting it here you send the message that a dispatch thing is happening with the type parameter, when actually there is none. That said, if you envision the possibility to widen the accepted type parameters for `AbsSeriesRing`, for which this definition of `characteristic` would be wrong, then it could make sense to keep the type constraint on `T` in this method signature. Otherwise, no.

Reply:
If you are interested, there is a couple of corner cases where adding a `where` clause is different, as it forces specialization (so it's still semantically equivalent): cf. https://docs.julialang.org/en/v1.4-dev/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing-1

Reply:
I don't mean where or not. I meant being more explicit about the types in the signatures.

And yes we also want specialization on all types.

Reply:
There are things like group rings and so on where you have group elements instead of ring elements. I'd prefer to keep the restriction, as this code is only valid when there is a ring. Besides that, we have done this throughout much of AbstractAlgebra and Nemo, and it would be major diversion to fix it everywhere.

Also, we do use it for documentation purposes. We try to keep the docstrings and the function prototypes the same. This way the reader of the documentation sees the restriction explicitly.

Maybe one day we decide to change our minds on this. But for now, that's how it's been done.

Reply:
> And yes we also want specialization on all types.

Sure, but my remark on specialization doesn't apply in this case, I mention it just FYI.

> I don't mean where or not. I meant being more explicit about the types in the signatures.

The topic was `where` clauses, so I'm not sure what you mean. But I would like to know what is the purpose of repeating the type constraints in all methods applying to this type when the type constraint is already on the type itself. It's a Julia anti-pattern.

Reply:
> We try to keep the docstrings and the function prototypes the same

It should be also removed from the docstrings, it's a useless load on the user, and hurts the eye of some julia developers.
It doesn't have to be fixed all at once.

Reply:
Because if we ever lift that restriction on the type, we'd need to go through and add precisely all the restrictions that are now there to be sure that it still works the way that was intended by the people who wrote the functions.

I recall Jeff Bezanson complaining about us doing this, very early on in the development of Nemo. We simply disagreed and have quite often written the restriction on the functions.

I agree that it is more typing and it makes it look more cluttered. But for mathematics, it isn't an awful choice.

Anyhow, I'm not going to go through and change the million or so such cases right now, and that would be orthogonal to this PR. So let's leave it for now. Maybe one day we change our minds. But it's certainly been a deliberate choice on our part.

Reply:
The difference is we are writing a maths library, not a library of container types.

Reply:
> Because if we ever lift that restriction on the type, we'd need to go through and add precisely all the restrictions that are now there to be sure that it still works the way that was intended by the people who wrote the functions.

It's not a clear win: without these constraints on the methods, you might want to go through all the methods to be sure that they are still valid, and you might have to add constraints on a few of them. With the current setting, you would probably have to update most of them, as they work equally well with the new possible type parameters. 

> I'm not going to go through and change the million or so such cases right now,

I certainly didn't request that. I would hope that the bar to make a cosmetic change like that isn't to change everything all at once. It's then very discouraging.

Otherwise sure, it's orthogonal to this PR.

> But it's certainly been a deliberate choice on our part.

Ok, it sounded otherwise in your previous comment, where you had concerns about covariance and stuff.

Reply:
Of course that has to be the bar. If you change it in some places and not others, people will use examples where it hasn't been changed as an excuse to not change their own code.

And making that the bar prevents people from wasting enormous quantities of time on things that have already been decided long ago because they personally think it should look different.

As I've said before, when I work on a project, I always try to adopt precisely the same conventions as the people who maintain the project, even if they are not the conventions I like. I don't try to change the conventions because I personally think they should have picked a different convention.

Of course, suggesting that there is an improvement to be made is fine. But I think you agree, changing this everywhere would be a massive waste of time, especially given the time constraints we have.

The priority has to be given to fixing things that close actual open tickets, not making cosmetic changes to conventions because of personal preferences!

Reply:
I have made clear that I didn't suggest to change everything, and I initiated the discussion because the reason for the `where` was not clear to me, I though it was an oversight. You don't need to waste more time telling me that we must not spend huge amount of time on trivialities, I have already completely understood this and this is not my intention. Nor that I should be more respectful of the conventions of a project. I generally try to follow them and to be a "good Open Source contributor" (to use your terms). But I can't stay silent if I believe something is a mistake.

Reply:
> Of course that has to be the bar.

The problem with this approach is that then: just assuming for one moment that you agree with the proposed change here for the sake of the argument, then the change could be made only eventually, when we can afford to spend some time on it. But for consistency, we keep the current way for new code, which means that we increase the work which has to be done eventually. So of course, consistency is nice, but for cosmetic changes like that it has a price that I think is too high to pay.

Reply:
We can hire a HiWi to do stuff like this. It's a complete waste of developer effort. That's assuming we agreed to the change, which I don't at present. And unless I misunderstand, nor does @thofma

We seriously need to focus on the stuff we know is broken and for which there is a clear benefit to fix. There is so much of it we are literally drowning in it at present.

Your argument about keeping the current way only increasing work later is absolutely valid, *if* we had made a decision to make the change, which we have not. But once we make that decision, we would put someone on the task and get it all done. Simply starting and not finishing hundreds of changes which we'd eventually like to have will leave the entire library in an unmaintainable mess. We must value consistency where it is possible.

Reply:
Anyway, let's not talk hypotheticals. On the way home I tried to think through what might happen in the future. I thought of four different possibilities:

1) Extension of the current RingElement type to something more general. We already have non-commutative rings, for which polynomials for example, have their own special type NCPoly (IIRC). But it is certainly feasible that we decide to have polynomials or series or some other thing that currently depends on coefficients being in a commutative ring that we want to lift that restriction for.

2) There are multiple packages out there that depend on our type hierarchy. Perhaps one of the packages needs to define polynomials over their special types which do not satisfy our RingElement interface, and the easiest way to do it is to extend the type to a union of our RingElement and their type (or some abstract type they create in AbstractAlgebra for this purpose). Then they go through and modify the handful of functions that won't work with their funky type.

3) Inexact rings can get really problematic. We might eventually decide to push them outside RingElement for some reason.

4) In implementing some bizarre mathematical structure, we decide the easiest way is to allow polynomials or series or some other type to accept something other than RingElement's for coefficients, and to build the new type around such extended polynomials, ensuring that the various functions required for that are extended to accept that new type.

Those are the three potential cases I can think of. I can't see any clearer than anyone else how likely any of these are, or how much of a burden it would be in either direction to do this sort of thing. But the fact remains that in mathematics, specifying exactly which kinds of objects are acceptable is fairly important. Simply accepting Any really jars my mathematical intuition, even if the type itself is currently so restricted.

Reply:
Having said all that, I do agree that the current convention Julia uses is correct for Julia. Many more of their objects are generic containers, and simply restricting things unnecessarily usually just leads to users complaining they can't do something that they otherwise could do if it weren't so restricted. This is a case Jeff Bezanson has made, and I agree with him on it, for the specific case of a general purpose language such as Julia.